Indices

aaronjridley opened the pull request:
# Description

Created a method for reading in indices, such as F10.7 and IMF. Users can now use this method for reading in new indices and then allowing users to gain access to the indices.

Addresses #12

There are a bunch of changes because I did this badly.  I added the ability to add indices and stuff, but then I made a bunch of code modifications in this which rippled through the whole code.  For example, I created the aether.h file, which includes all of the other include files.  This caused problems with the inter-dependencies between the different include files, which caused me to have to move things. For example, some EUV stuff can be argued to be either EUV or Neutrals, so it used to be in neutrals, but then there were neutral dependencies in EUV and EUV dependencies in Neutrals. So, I moved this stuff to EUV. Basically, I cleaned up the code a bit in this process.

## Type of change


- New feature (non-breaking change that adds functionality): this is the indices addition.
- Bug Fixes: There were some bugs in the SZA calculation and the ionization initialization. These were fixed. In order to fix them, I then changed how the grid is created, to allow very simple creation using the aether.in file.
- Breaking change (fix or feature that would cause existing functionality to not work as expected): this is the modification of the include files, which caused things to break, which was then fixed.
- This change requires a documentation update: Ugh, I really need to do some documentation!


# How Has This Been Tested?

- Running the code over and over and over again, making plots, and seeing of things look correct.  I output the SZA now, so this can be explored. 
- A new plotting code was developed and added. 

## Test configuration

* Operating system: Linux (PopOS)
* Compiler, version number: gcc 9.3.0
* I don't think that I have anything special.

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [Well. I could do better.] I have commented my code, particularly in hard-to-understand areas
- [Yeah, we need to work on this.] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [Still need to do this.] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [N/A] Add a note to ``CHANGELOG.md``, summarizing the changes

aburrell commented:
Hi @aaronjridley I edited your above comment by:

1) removing the space between the hash symbol and the issue number so it hyper-links, and
2) ticking the "merge into develop" box.  Because you did, and it is applicable.

Taking a look at the code now.

aburrell commented on CONTRIBUTING.md:
````diff
@@ -29,6 +29,8 @@ python3 -m pip cpplint
 python3 -m pip --user cpplint
 ```
 
+Using a linter in an editor is good too, although it has been found that these are not as good.  The linter in the 'atom' editor is ok, but you have to install the packages `linter` and `gcc-linter`. In atom, you can install the packages `whitespaces` and `tabs-to-spaces` to automatically remove whitespaces at the end of the lines, and convert tabs to spaces.
````
```suggestion
Using a linter in an editor is another good option, although its may have employed different linting standards.  It may also require that you install additional packages.  For example, the linter in the `atom` editor depends on the `linter` and `gcc-linter` packages for basic linting and the `whitespaces` and `tabs-to-spaces` packages to automatically remove whitespaces at the end of the lines, and convert tabs to spaces.
```

aburrell replied:
Needs to be broken up so that maximum line length is 79 or 80 (don't recall what we decided on).

aburrell commented on include/bfield.h:
```diff
@@ -4,24 +4,21 @@
-#include "../include/planets.h"
-#include "../include/inputs.h"
-#include "../include/report.h"
+// The armadillo library is to allow the use of 3d cubes and other
+// array types, with array math built in. This eliminates loops!
```
Is this comment relevant here?

aaronjridley replied:
Done

aburrell commented on include/indices.h:
```diff
@@ -18,25 +31,57 @@ class Indices {
+  std::vector<index_time_pair> all_indices_arrays;
 
-  float get_index(double time, std::vector<ind_time_pair> index);
+  int f107_ = 0;
+  int f107a_ = 1;
+  int imf_bx_ = 2;
+  int imf_by_ = 3;
+  int imf_bz_ = 4;
+  int sw_vx_ = 5;
+  int sw_vy_ = 6;
+  int sw_vz_ = 7;
+  int sw_n_ = 8;
+  int sw_t_ = 9;
+  int ae_ = 10;
+  int al_ = 11;
+  int au_ = 12;
```
Just a suggestion!  It looks like these are indices for particular values.  Perhaps we could adopt a convention for indices throughout the code?  Potential examples:
- `ind_f107_`
- `i_f107_`
- `if107_`

aburrell replied:
> which is weird, since I always felt like you should.

Same here.  I think it's silly to not identify an index when it could easily be mistaken for a value based on the variable name.  That makes it a non-descriptive name, in my book!

1) Would we employ CamelCase for these variables to make it easier? -> As long as we're not in Python, absolutely!

2) I am not the right person to answer this one, I was hoping to pick up the reason behind the trailing underscore by context, but haven't gotten there yet 😋

aaronjridley replied:
Ok, Qusai's comments are below this, but I will respond here.....

I was debating about how to implement this in the best way.  I am not sure that I came up with the "best" way, but the method that I did come up with has these indices being constants.  So, I should change them to "const int".

The next thought on naming is somewhat important. Let's take two examples (IMF Bz and F107a).  What should we name these indices? Summing up suggestions, we have:
1. iImfBz_ and iF107a_
2. i_imf_bz_ and i_f107a_
3. IMFBZ and F107A
4. _IMFBZ and _F107A
5. iIMFBZ and iF107A
6. iIMFBZ_ and iF107A_

I prefer 6, since it gives all sorts of indication that it is an index and a constant.

How would/should we do physical constants then? Like pi and dotr (degrees to radians) and boltzmanns_constant? Are these supposed to be all CAPS? If so, then we need to come up with better names for boltzmanns_constant and the like.

aburrell replied:
I like 5 and 6. Boltsmann's constant is $k_B$, so maybe `KB` or `KBOLTS` if you're worried about people forgetting all of the constants we're supposed to remember forever.

aburrell replied:
I am fine with that, just create an issue and link it in a comment here to resolve the conversation.

aaronjridley replied:
Ok - issue #44 is now linked to this.  I probably did this incorrectly, since that issue is linked to this whole pull request, and may disappear when this is dealt with.  So, I am unsure of how to do this properly.

aburrell replied:
I think you did it fine, GitHub is pretty good about leaving Issues open unless you use closing words just right.

aburrell reviewed (commented):
Partial review (11/55 files).

qalshidi commented:
If possible can the `_` be before the name and not after? Keeping in mind auto-complete it would suck if I'm trying to autocomplete `f107_init` or something and this shows up before. These are also indeces which means they are constant values (not to be confused with physical constants). Shouldn't they `const int F107 = 0; const int F107A = 1;` etc? Isn't that the C++ naming convention? I would prefer not to see CamelCase just for an edge case.

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+import numpy as np
+from netCDF4 import Dataset
+
+#-----------------------------------------------------------------------------
+# Get some ancillary information from the netcdf file.  This simply
+# mirrors the gitm header info.
+# Inputs:
+#  - filelist: a list of netcdf files. Really, only the last file is used.
+#              also, the number of files is recorded.
+# Output:
+#  - header: A dictionary containing information about the netcdf file, such
+#            as nLons, nLons, nAlts, nVars, variable names, time(s)
+#-----------------------------------------------------------------------------
```
This needs to be in the function docstring.  Here's an example: https://numpydoc.readthedocs.io/en/latest/example.html

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+
+def read_aether_header(filelist):
+
+    header = {}
+    header["nFiles"] = len(filelist)
+    header["version"] = 0.1
+    header["nLons"] = 0
+    header["nLats"] = 0
+    header["nAlts"] = 0
+    header["nVars"] = 0
+    header["vars"] = []
+    header["time"] = []
+    header["filename"] = []
+
+    header["filename"].append(filelist[-1])
```
```suggestion
    header = {"nFiles": len(filelist), "version": 0.1, "nLons": 0, "nLats": 0, "nAlts": 0, "nVars": 0: "vars": [], "time": [], "filename": [filelist[-1]]}
```

aaronjridley replied:
Done

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+
+    header["filename"].append(filelist[-1])
+    
+    ncfile = Dataset(filelist[-1],'r')
```
```suggestion
    ncfile = Dataset(filelist[-1], 'r')
```

Looks like this needs to be linted.

aburrell replied:
Also, if you use `with`, file closure happens cleanly even if the code errors.

aaronjridley replied:
I have not used "with" and am not sure how to use it.  I will investigate this.  My guess is that this python code will not be super important as others develop better ones....  This is just sort of a start.

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+            header["nLats"] = s[1]
+            header["nAlts"] = s[2]
+            header["vars"].append(var.name)
+            header["nVars"] = header["nVars"] + 1
```
```suggestion
            header["nVars"] += 1
```

aaronjridley replied:
Done!

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+            header["vars"].append(var.name)
+            header["nVars"] = header["nVars"] + 1
+
+    base = datetime(1965,1,1,0,0,0)
```
```suggestion
    base = datetime(1965, 1, 1, 0, 0, 0)
```

aaronjridley replied:
Done!

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+
+    base = datetime(1965,1,1,0,0,0)
+    time = np.array(ncfile.variables['Time'])
+    current = base + timedelta(seconds = time[0])
```
```suggestion
    current = base + timedelta(seconds=time[0])
```

aaronjridley replied:
I didn't make this change.  I don't know why to NOT have spaces here.  Maybe because it is a parameter?  I like the extra spaces.

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+    current = base + timedelta(seconds = time[0])
+    data["time"] = current
+
+    iVar = 0
+    for var in vars:
+        data[iVar] = np.array(ncfile.variables[var])
+        iVar = iVar + 1
```
```suggestion
    for i, var in enumerate(vars):
        data[i] = np.array(ncfile.variables[var])
```

aaronjridley replied:
Done like this but not with variable named i, but iVar

aburrell replied:
In Python CamelCase is used to denote class names.  Since this isn't a class it needs to be lower or snake case.

aburrell reviewed (changes requested):
Another partial review.  Python comments should be applied to other python files.  Re-request a review from me once they're done!

aaronjridley commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+
+from datetime import datetime
+from datetime import timedelta
```
Done!

aaronjridley commented on python/aether_routines.py:
```diff
@@ -0,0 +1,84 @@
+from datetime import datetime
+from datetime import timedelta
+import numpy as np
+from netCDF4 import Dataset
```
Didn't know that this was a thing!!! Done!

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,96 @@
+              "time": [], \
+              "filename": [filelist[-1]] }
+    
+    ncfile = Dataset(filelist[-1], 'r')
+    header["nVars"] = 0
+    for var in ncfile.variables.values():
+        s = var.shape
+        if (len(s) == 3):
+            header["nLons"] = s[0]
+            header["nLats"] = s[1]
+            header["nAlts"] = s[2]
+            header["vars"].append(var.name)
+            header["nVars"] += 1
+
+    base = dt.datetime(1965, 1, 1, 0, 0, 0)
+    time = np.array(ncfile.variables['Time'])
+    current = base + dt.timedelta(seconds = time[0])
+    header["time"].append(current)
+            
+    ncfile.close()
```
```suggestion
    with Dataset(filelist[-1], 'r') as ncfile:
        header["nVars"] = 0
        for var in ncfile.variables.values():
            if (len(var.shape) == 3):
                header["nLons"] = var.shape[0]
                header["nLats"] = var.shape[1]
                header["nAlts"] = var.shape[2]
                header["vars"].append(var.name)
                header["nVars"] += 1

        base = dt.datetime(1965, 1, 1, 0, 0, 0)
        time = np.array(ncfile.variables['Time'])
        current = base + dt.timedelta(seconds=time[0])
        header["time"].append(current)
```

Implemented `with`.  Also removed definition of unnecessary variable.
The lack of whitespace for kwargs is useful for visually separating variable equalities from function or method kwarg input.

aburrell replied:
Question: is `time` ever going to be a list of more than one item?  I can't see how :(

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,96 @@
+#-----------------------------------------------------------------------------
+
+def read_aether_one_file(file, vars):
+    r""" Read in list of variables from a single netCDF file
+
+    Parameters
+    ----------
+    file: netcdf file to read
+    vars: list of variable NAMES to read
+
+    Returns
+    -------
+    data["time"]: datetime of the file
+    data[NUMBER]: data that is read in.
+                  NUMBER goes from 0 - number of vars read in (0-3 typical)
+
+    """
```
```suggestion
    """ Read in list of variables from a single netCDF file

    Parameters
    ----------
    file : str
        Name of netCDF file to read
    vars : list
        List of desired variable names to read

    Returns
    -------
    data : dict
        Dict with keys 'time', which contains a datetime object specifying the time of
        the file and indices ranging from 0 to `len(vars) - 1`, corresponding to the
        variable names in `vars` that holds arrays of the specified dat
    
    """
```

Why is an index used instead of the variable name?  It's an unusual approach.

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,96 @@
+
+    """
+
+    data = {}
+    
+    ncfile = Dataset(file, 'r')
+
+    base = dt.datetime(1965, 1, 1, 0, 0, 0)
+    time = np.array(ncfile.variables['Time'])
+    current = base + dt.timedelta(seconds = time[0])
+    data["time"] = current
+
+    for iVar, var in enumerate(vars):
+        data[iVar] = np.array(ncfile.variables[var])
+    ncfile.close()
+
+    return data
```
Also, why not just use the variable names as keys?  It would make the data dict more user friendly.

aburrell commented on src/Makefile.Ubuntu:
```diff
@@ -0,0 +1,96 @@
+	${AR} ${MY_LIB} ${CLASSES} ${OBJECTS}
+
+
+Aether: ${MAIN} LIB
+	${LINK.CPP} -o aether.exe ${MAIN} ${MY_LIB} -L../lib -L${LIB_DIR} -l${NETCDF}
+
+test: ${TEST} LIB
+	${LINK.CPP} -o test.exe ${TEST} ${MY_LIB} -L../lib
+
+clean:
```
Do we also want to have `all` and `install`?

aburrell commented on src/bfield.cpp:
```diff
@@ -45,3 +46,41 @@ bfield_info_type get_bfield(float lon,
+      lonlat(1) = -pi/2.0;
+    }
+  } else if (input.get_bfield_type() == "dipole") {
+    // This is an approximation right now, due to the fact that the
+    // pole on some planets (including Earth), have an offset, so the
+    // pole location should be altitude dependent.  For many planets,
+    // this is a very small error, but should be fixed at some point.
```
Is this captured in an issue?

aburrell replied:
I created #43 to resolve this.

aburrell commented on src/indices.cpp:
```diff
@@ -6,34 +6,99 @@
-  int iErr;
-  std::string file;
+// ----------------------------------------------------------------------
+//
```
What is supposed to go in this comment block?

aburrell commented on src/indices.cpp:
```diff
@@ -6,34 +6,99 @@
+          file_contents.times,
+          file_contents.values[iVar],
+          file_contents.missing_values[iVar]);
+  }  // if
```
I am a bit surprised that the `if` closure is indented less than the surrounding loops.

aburrell reviewed (changes requested):
Making progress!  I made some suggestions to the `aether_routines.py` that should help you finish cleaning up the rest of the python routines.   Let me know if you have questions/concerns about it.

There are also some questions and small areas for improved comments/functionality throughout.

aburrell commented on python/aether_routines.py:
```diff
@@ -0,0 +1,97 @@
+
+    return header
+    
+#-----------------------------------------------------------------------------
+# Read in some variables from one file. Inputs:
+#   - file: netcdf file to read
+#   - vars: list of variable NAMES to read
+# Output:
+#   - data["time"]: datetime of the file
+#   - data[NUMBER]: data that is read in.
+#                   NUMBER does from 0 - number of vars read in (0-3 typical)
+#-----------------------------------------------------------------------------
```
```suggestion
```

aburrell commented on src/inputs.cpp:
```diff
@@ -102,6 +103,14 @@ float Inputs::get_dt_euv() {
+}
+
+// -----------------------------------------------------------------------
+//
```
As discussed in email, this can be dealt with by:
1) Adding function name and a very short description
2) Adding a description to the contributing guidelines that says what the C++ function descriptions should look like

aburrell reviewed (commented):
So, there are a few issues that haven't been dealt with yet.  As per our email discussion, the problems with the python code can either be dealt with here, or put into an issue.  You can even assign the issue to me.

I don't think the function names and brief descriptions in the C++ code should be put off any further, because I don't want someone to start trying to contribute and not having a function description available.

aaronjridley commented:
I created an issue (I think #42) to create/modify the python code to be better written, have more features, and be more robust.

aaronjridley commented:
Ok, I don't see where to request a re-review. So, I am just adding a comment.  I really don't see how to do some of this stuff. Last night and this morning, I did the following things:

- Added comments above all functions, giving a short description of what they do (Somewhere in the comments above)
- Modified the indices constants, as we agreed upon. I did this because every single file was changed due to the comments, so I just added this in (Issue #41)
- Modified all physical constants in the code, as well as conversion constants (like degrees to radians and time conversions) (Issue #41)
- Made the change for Neutrals.neutrals[iSpecies] to Neutrals.species[iSpecies].  This makes it consistent with the Ions.species[iSpecies] schema and just makes more sense.  I did this because it has been on my list of to-do items for months, and I have been waiting to get a clean branch, but that isn't happening.  So, as long as we are changing all the files in the code, might as well make this relatively small change. (Issue #30 )
- Modified the README.md file to be more consistent with the way the code works now (src/Makefile has an include now, plotting code has changed).
- Modified to inputs/aether.in to be more consistent with the way the code works now (grid specifically)
- Added plots to inputs to allow users to compare the results of their tests. 

So, I don't really know what I am supposed to do now in term of requesting a re-review.

aburrell commented on src/planets.cpp:
```diff
@@ -165,23 +169,22 @@ int Planets::update(Times time) {
+    float y_heliocentric = sma*sqrt(1-ecc*ecc)*sin(ecc_anomaly * cDtoR);
 
-    float true_anomaly = atan2(y_heliocentric, x_heliocentric)*rtod;
+    //float true_anomaly = atan2(y_heliocentric, x_heliocentric)*rtod;
```
```suggestion
```

Suggest removing old code

aaronjridley replied:
Uncommented, since doing the calculation doesn't change anything.  I would like to include this, in case it is need by someone else, like a planetary person.

aburrell commented on src/neutrals.cpp:
```diff
@@ -309,68 +313,14 @@ void Neutrals::set_bcs(Report &report) {
   // temperature_scgc.slice(nAlts-2).fill(800.0);
   // temperature_scgc.slice(nAlts-1).fill(800.0);
```
@aaronjridley I can't make a commitable suggestion here, but also recommend deleting lines 311 and 312.

aburrell reviewed (commented):
Two remaining suggestions.

I haven't fixed my C++ setup yet, so am tagging @Ionospheres to make sure this PR compiles and runs.

Ionospheres commented:
While the code seems to compile just fine, I'm getting an instant seg fault when I try to run it. I'm going to keep troubleshooting...

Ionospheres reviewed (commented):
The code compiles and runs nicely for me now. Outputs look reasonable for the physics included.

aburrell commented:
@Ionospheres, does that mean you approve?  If so, go ahead and leave an approving review 😄

Ionospheres reviewed (approved):
Approving review (and I found the right github button!) 😄

aaronjridley commented:
Yeah!  I am pulling the trigger on this.
